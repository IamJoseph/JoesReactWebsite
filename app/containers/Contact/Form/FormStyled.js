import styled, { keyframes } from 'styled-components';

/*
 * Form Elements
 */

export const Form = styled.form`
  background: linear-gradient(to bottom, #fffef6 50%,#edebe1 90%, #edebe1 100%);
  box-shadow: 0px 5px 9px -2px #ab9393;
  display: flex;
  flex-direction: column;
  margin: 0 auto 20px auto;
  max-width: 800px;
  padding: 1em;
  width: 100%;
`;

export const Input = styled.input`
  border: none;
  width: 100%;
`;

export const FieldSet = styled.fieldset`
  border-radius: 3px;
  margin: .25em 2px;
  width: ${({ halfWidth }) => halfWidth ? '50%' : '100%'};
`;

export const TextArea = styled.textarea`
  border: none;
  width: 100%;
`;

export const SubmitButton = styled.button`
  align-self: flex-end;;
  background-color: #81a5c3;
  border: none;
  border-bottom: 1.5px solid #6c8499;
  border-top: 1.5px solid #8fb3d1;
  border-radius: 10px;
  color: #fffef6;
  display: flex;
  justify-content: center;
  height: 3.343em;
  outline: none;
  width: 6em;
  &.loading, &:focus{
    background-color: #b7c8d2;
    border-bottom: 1.5px solid #b7c8d2;
    border-top: 1.5px solid #b7c8d2;
  }
`;

/*
 * Spinner
 */

const colors = keyframes`
 0% { stroke: #4285F4; }
 25% { stroke: #DE3E35; }
 50% { stroke: #F7C223; }
 75% { stroke: #1B9A59; }
 100% { stroke: #4285F4; }
`;

const dash = keyframes`
 0% { stroke-dashoffset: 187; }
 50% {
   stroke-dashoffset: 46.75;
   transform:rotate(135deg);
 }
 100% {
   stroke-dashoffset: 187;
   transform:rotate(450deg);
 }
`;

const rotator = keyframes`
 0% { transform: rotate(0deg); }
 100% { transform: rotate(270deg); }
`;

export const Spinner = styled.svg`
 animation: ${rotator} 1.4s linear infinite;
 display: flex;
 align-items: center;
 justify-content: center;
 width: 100%;
 height: 100%;
`;

export const Circle = styled.circle`
 stroke-dasharray: 187;
 stroke-dashoffset: 0;
 transform-origin: center;
 animation: ${dash} 1.4s ease-in-out infinite, ${colors} 5.6s ease-in-out infinite;
`;
