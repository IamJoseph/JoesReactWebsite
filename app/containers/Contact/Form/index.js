import React from 'react';
import PropTypes from 'prop-types';
import { Form, FieldSet, Input, TextArea, SubmitButton, Spinner, Circle } from './FormStyled';

export default function FormComponent(props) {
  const { handleSubmit, handleChange, loading, name, message, email } = props;

  const submitHandler = () => {
    if (loading) {
      return (
        <Spinner viewBox="0 0 66 66">
          <Circle
            fill="none"
            strokeWidth="6"
            strokeLinecap="round"
            cx="33"
            cy="33"
            r="30"
          />
        </Spinner>
      );
    }
    return <div style={{ width: '100%' }}>Send</div>;
  };

  return (
    <Form
      method="POST"
      onSubmit={handleSubmit}
      encType="text/plain"
    >
      <label htmlFor="name">Name: </label>
      <FieldSet halfWidth>
        <Input
          id="name"
          name="name"
          placeholder="Optimus Prime"
          required
          value={name}
          onChange={(e) => handleChange(e, 'name')}
        />
      </FieldSet>
      <label htmlFor="message">Message: </label>
      <FieldSet>
        <TextArea
          id="message"
          name="message"
          required
          rows="10"
          value={message}
          placeholder="Your message"
          onChange={(e) => handleChange(e, 'message')}
        >
        </TextArea>
      </FieldSet>
      <label htmlFor="email"><em>Your</em> Email Address: </label>
      <FieldSet halfWidth>
        <Input
          id="email"
          name="email"
          type="email"
          value={email}
          required
          placeholder="your.name@email.com"
          onChange={(e) => handleChange(e, 'email')}
        />
      </FieldSet>
      <SubmitButton
        className={loading && 'loading'}
        onClick={() => { document.activeElement.blur(); }}
      >
        { submitHandler() }
      </SubmitButton>
    </Form>
  );
}

FormComponent.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  name: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
};

FormComponent.defaultProps = {
  loading: false,
  name: '',
  message: '',
  email: '',
};
