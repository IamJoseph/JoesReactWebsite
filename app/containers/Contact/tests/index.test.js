import React from 'react';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import toJson from 'enzyme-to-json';
import { Contact } from '../index';
import Form from '../Form';

let fakeProps = {
  dispatch: () => {},
};
describe('<Contact {...fakeProps} />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Contact {...fakeProps} />);
    wrapper.setProps({ success: null });
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ success: true });
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ success: false });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should update state with changes via handleChange', () => {
    const wrapper = mount(<Contact {...fakeProps} />);
    wrapper.find(Form).find('textarea').simulate('change', { target: { value: 'This is a message' } });
    expect(wrapper.state('message')).toBe('This is a message');
  });

  it('handles the form submit', () => {
    const spy = jest.fn();
    const wrapper = mount(<Contact dispatch={spy} />);
    wrapper.setState({
      name: 'me',
      message: 'message',
      email: 'ok@whatever.com',
      loading: false,
    });
    wrapper.find(Form).simulate('submit');
    expect(spy).toHaveBeenCalled();
  });

  it('does not submit when the form is incomplete', () => {
    const spy = jest.fn();
    const wrapper = mount(<Contact dispatch={spy} />);
    wrapper.setState({
      name: 'me',
      message: '',
      email: 'ok@whatever.com',
      loading: false,
    });
    wrapper.find(Form).simulate('submit');
    expect(spy).not.toHaveBeenCalled();
  });

  it('calls componentWillReceiveProps and sets state', () => {
    const wrapper = mount(
      <Contact {...fakeProps} />
    );
    const fakeInitialState = {
      name: '',
      message: '',
      email: '',
      loading: false,
    };
    const spy = sinon.spy(Contact.prototype, 'componentWillReceiveProps');
    expect(Contact.prototype.componentWillReceiveProps.called).toBe(false);
    wrapper.setProps({ fake: 'prop' });
    expect(Contact.prototype.componentWillReceiveProps.called).toBe(true);
    expect(wrapper.state()).toEqual(fakeInitialState);
    spy.reset();
  });

  it('should call handleSubmit when form is submited', () => {
    fakeProps = {
      handleChange: () => {},
    };
    const onSubmitSpy = jest.fn();
    const formComponent = mount(<Form {...fakeProps} handleSubmit={onSubmitSpy} />);
    formComponent.simulate('submit');
    expect(onSubmitSpy).toHaveBeenCalled();
  });
});
