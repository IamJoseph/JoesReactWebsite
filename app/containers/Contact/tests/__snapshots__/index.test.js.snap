// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contact {...fakeProps} /> renders correctly 1`] = `
<styled.section>
  <HelmetWrapper
    encodeSpecialCharacters={true}
  >
    <title>
      Contact
    </title>
    <meta
      content="A contact form and different social media methods to contact Joseph Ireland"
      name="description"
    />
  </HelmetWrapper>
  <Links />
  <styled.div>
    <styled.h1>
      Send an Email!
    </styled.h1>
  </styled.div>
  <FormComponent
    email=""
    handleChange={[Function]}
    handleSubmit={[Function]}
    loading={false}
    message=""
    name=""
  />
</styled.section>
`;

exports[`<Contact {...fakeProps} /> renders correctly 2`] = `
<styled.section>
  <HelmetWrapper
    encodeSpecialCharacters={true}
  >
    <title>
      Contact
    </title>
    <meta
      content="A contact form and different social media methods to contact Joseph Ireland"
      name="description"
    />
  </HelmetWrapper>
  <Links />
  <styled.div>
    <styled.h1>
      Thanks, I'll Get Back to You Soon :)
    </styled.h1>
  </styled.div>
  <FormComponent
    email=""
    handleChange={[Function]}
    handleSubmit={[Function]}
    loading={false}
    message=""
    name=""
  />
</styled.section>
`;

exports[`<Contact {...fakeProps} /> renders correctly 3`] = `
<styled.section>
  <HelmetWrapper
    encodeSpecialCharacters={true}
  >
    <title>
      Contact
    </title>
    <meta
      content="A contact form and different social media methods to contact Joseph Ireland"
      name="description"
    />
  </HelmetWrapper>
  <Links />
  <styled.div>
    <styled.h1>
      Sorry There Was an Error :( Try again later?
    </styled.h1>
  </styled.div>
  <FormComponent
    email=""
    handleChange={[Function]}
    handleSubmit={[Function]}
    loading={false}
    message=""
    name=""
  />
</styled.section>
`;
